File: C++/LIS.cpp
#include<iostream>
#include<vector>
#include<algorithm>

class Solution{
    public:
    int LIS(std::vector<int>& arr) {
        int n =arr.size();
        if(n==0){return 0;}
        std::vector<int>dp(n,1);
        int maxLength =1;
        for(int i =1;i<n;i++){
            for(int j=0;j<i;j++){
                if(arr[j]<arr[i]){
                    dp[i]=std::max(dp[j]+1,dp[i]);
                }
            }
            maxLength = std::max(maxLength, dp[i]);
        }
        return maxLength;
    }
};

int main(){
    std::vector<int>arr={6,3,1,5,2,3,7};
    Solution solution;
    std::cout<<solution.LIS(arr)<<std::endl;
    return 0;
}

File: python/LIS.py
class Solution:
    def LIS(self , arr: list[int]) -> int:
        n=len(arr)
        if n==0:
            return 0
        dp=[1]* n
        max_length=1
        for i in range(1,n):
            for j in range(0,i):
                if arr[j]<arr[i]:
                    dp[i]=max(dp[i],dp[j]+1)
            max_length=max(max_length,dp[i])
        return max_length
    
arr=[6,3,1,5,2,3,7]
s=Solution()
print(s.LIS(arr))